volatility3/framework/plugins/linux/mountinfo.py

```diff
@@ -0,0 +1,227 @@
+            mount_ns_id = mnt_namespace.get_inode()
+            
+            if self._show_mountpoints_per_namespace:
+                if mount_ns_id in seen_namespaces:
+                    continue
+                else:
+                    seen_namespaces.add(mount_ns_id)
+            
+            for mount in self._get_mnt_namespace_mountpoints(mnt_namespace):
+                yield task, mount, mount_ns_id
+
+    def _generator(self):
+        pids = self.config.get('pids')
+
+        for task, mnt, mnt_ns_id in self._get_tasks_mountpoints(pids=pids):
+            mnt_info = self.get_mountinfo(mnt, task)
+            if mnt_info is None:
+                continue
+
+            if self.config.get('mount-format'):
+                all_opts = set()
+                all_opts.update(mnt_info.mnt_opts)
+                all_opts.update(mnt_info.sb_opts)
+                all_opts_str = ",".join(all_opts)
+
+                extra_fields_values = [mnt_info.devname, mnt_info.path_root, mnt_info.mnt_type, all_opts_str]    
+            else:
+                mnt_opts_str = ",".join(mnt_info.mnt_opts)
+                fields_str = " ".join(mnt_info.fields)
+                sb_opts_str = ",".join(mnt_info.sb_opts)
+
+                extra_fields_values = [mnt_info.mnt_id, mnt_info.parent_id, mnt_info.st_dev, mnt_info.mnt_root_path, 
+                                       mnt_info.path_root, mnt_opts_str, fields_str,  mnt_info.mnt_type, 
+                                       mnt_info.devname, sb_opts_str]
+            
+            fields_values = [mnt_ns_id]
+            if not self._show_mountpoints_per_namespace:
+                fields_values.append(task.pid)
+            fields_values.extend(extra_fields_values)
+            
+            yield (0, fields_values)
+
+    def run(self):
+        if self.config.get('all-processes') and self.config.get('pids'):
+            raise ValueError("Unable to use --all-processes and specified a pid")
+        
+        # When no arguments are specified, it displays the mountpoints per namespace
+        self._show_mountpoints_per_namespace = not any([self.config.get('pids'), self.config.get('all-processes')])
+
+        columns = [("MNT_NS_ID", int)]
+        if not self._show_mountpoints_per_namespace:
+            columns.append(("PID", int))
+        
+        if self.config.get('mount-format'):
+            extra_columns = [("DEVNAME", str), ("PATH", str), ("FSTYPE", str), ("MNT_OPTS", str)]
+        else:
+            # /proc/[pid]/mountinfo output format
+            extra_columns = [("MOUNT ID", int), ("PARENT_ID", int), ("MAJOR:MINOR", str), ("ROOT", str),
+                             ("MOUNT_POINT", str), ("MOUNT_OPTIONS", str), ("FIELDS", str), ("FSTYPE", str),
+                             ("MOUNT_SRC", str), ("SB_OPTIONS", str)]
+            
+        columns.extend(extra_columns)
+
+        return renderers.TreeGrid(columns, self._generator())
```

Comment:
This is kind of icky.  The number of generator can produce different numbers of columns based on state (`_show_mountpoints_per_namespace`) rather than a parameter, and the columns are separately set by the same state.  I'd be much happier with either:
* A system that always included the pid (again, may just be me not understanding in what circumstances you *wouldn't* had an associated pid)
* An if statement that had the two column lists and then provided a parameter to the generator, so that you could clearly see based on the parameter that you either get one type of output or another.

I'm thinking of this for other plugins that might want to use the information and them needing to figure out which form the plugin returns.  I'd also be tempted to just fill the Pid column with `NotApplicable` so that the shape of the output data doesn't change between runs/options...

Reply:
> The number of generator can produce different numbers of columns based on state (_show_mountpoints_per_namespace) rather than a parameter, and the columns are separately set by the same state. 

The `_show_mountpoints_per_namespace` is indeed the plugins parameters, It's just a simplification to avoid checking both arguments every single time.
`self._show_mountpoints_per_namespace = not any([self.config.get('pids'), self.config.get('all-processes')])`

The **--mount-format** completely changes the output view and make it looks like the userland _mount_ output or _/proc/pid/mounts_, which most Linux users are more familiar with. The _mountinfo_ output format (default), which is based on the _/proc/pid/mount_ output, has quite more details.
I made the layout of `run()` and the bottom of  `_generator()` as similar as possible to make it easier to modify in the future.

Reply:
If it's just for display then the plugin shouldn't be making that decision.  I know it's difficult and it feels like you're being unkind in the most common use of the plugin, but it's really not the plugin's job to decide what data it should and shouldn't return.

One of the main goals of volatility 3 was to allow it to be used as a library, and potentially allow the data to be processed automatically.  Given the UI might be a web UI, or a GUI, where you can toggle columns as you want, reorder them etc, it doesn't make sense for there to be a flag just to appease the CLI users.  I really don't want plugins trying to make decisions because of what they think will look nice for the CLI specifically.  It will limit people's thinking and set a bad precedent.

It *looks* as though removing the pid column is the only thing that changes, and in my opinion that's really not worthwhile.  For this to be used as a library, the plugin needs to return everything and then let whatever UI is in use make decisions about what's displayed and how.  We can try our best to make it easily human readable, but display stuff should really be dealt with by the UI rather than library or plugins.

Reply:
Fixed